loris/utils.py

```diff
@@ -41,3 +44,93 @@ def symlink(src, dst, force=True):
         os.unlink(dst)
 
     os.symlink(src, dst)
+
+
+def remove_f(path):
+    """Remove a file at ``path``, but don't error if it doesn't exist."""
+    try:
+        os.remove(path)
+    except OSError as err:
+        if err.errno == errno.ENOENT:
+            pass
+        else:
+            raise
+
+
+def safe_rename(src, dst):
+    """Rename a file from ``src`` to ``dst``.
+
+    We use a custom version rather than the standard library because we
+    have two requirements:
+
+    *   Moves must be atomic.  Otherwise Loris may serve a partial image from
+        a cache, which causes an error.  ``shutil.move()`` is not atomic.
+
+        Note that multiple threads may try to write to the cache at once,
+        so atomicity is required to ensure the serving on one thread doesn't
+        pick up a partially saved image from another thread.
+
+    *   Moves must work across filesystems.  Often temp directories and the
+        cache directories live on different filesystems.  ``os.rename()`` can
+        throw errors if run across filesystems.
+
+    So we try ``os.rename()``, but if we detect a cross-filesystem copy, we
+    switch to ``shutil.move()`` with some wrappers to make it atomic.
+    """
+    logger.debug('Renaming %r to %r', src, dst)
+    try:
+        os.rename(src, dst)
+    except OSError as err:
+        logger.debug('Calling os.rename(%r, %r) failed with %r', src, dst, err)
+        if err.errno == errno.EXDEV:
+
+            # This is based on the safe, atomic file copying algorithm
+            # described in https://stackoverflow.com/a/28090883/1558022.
```

Comment:
Isn't some of this unnecessary, because we don't care if the file gets overwritten? If one process creates dst atomically, and then another process overwrites it atomically, is there any harm?